File: android/uhabits-android/src/main/java/org/isoron/uhabits/widgets/BaseWidgetProvider.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.widgets;

import android.appwidget.*;
import android.content.*;
import android.os.*;
import androidx.annotation.*;
import android.util.*;
import android.widget.*;

import org.isoron.uhabits.*;
import org.isoron.uhabits.core.models.*;
import org.isoron.uhabits.core.preferences.*;

import java.util.*;

import static android.appwidget.AppWidgetManager.*;
import static org.isoron.androidbase.utils.InterfaceUtils.dpToPixels;

public abstract class BaseWidgetProvider extends AppWidgetProvider
{
    private HabitList habits;

    private Preferences preferences;

    private WidgetPreferences widgetPrefs;

    public static void updateAppWidget(@NonNull AppWidgetManager manager,
                                       @NonNull BaseWidget widget)
    {
        RemoteViews landscape = widget.getLandscapeRemoteViews();
        RemoteViews portrait = widget.getPortraitRemoteViews();
        RemoteViews views = new RemoteViews(landscape, portrait);
        manager.updateAppWidget(widget.getId(), views);
    }

    @NonNull
    public WidgetDimensions getDimensionsFromOptions(@NonNull Context ctx,
                                                     @NonNull Bundle options)
    {
        int maxWidth =
            (int) dpToPixels(ctx, options.getInt(OPTION_APPWIDGET_MAX_WIDTH));
        int maxHeight =
            (int) dpToPixels(ctx, options.getInt(OPTION_APPWIDGET_MAX_HEIGHT));
        int minWidth =
            (int) dpToPixels(ctx, options.getInt(OPTION_APPWIDGET_MIN_WIDTH));
        int minHeight =
            (int) dpToPixels(ctx, options.getInt(OPTION_APPWIDGET_MIN_HEIGHT));

        return new WidgetDimensions(minWidth, maxHeight, maxWidth, minHeight);
    }

    @Override
    public void onAppWidgetOptionsChanged(@Nullable Context context,
                                          @Nullable AppWidgetManager manager,
                                          int widgetId,
                                          @Nullable Bundle options)
    {
        try
        {
            if (context == null) throw new RuntimeException("context is null");
            if (manager == null) throw new RuntimeException("manager is null");
            if (options == null) throw new RuntimeException("options is null");
            updateDependencies(context);
            context.setTheme(R.style.WidgetTheme);

            BaseWidget widget = getWidgetFromId(context, widgetId);
            WidgetDimensions dims = getDimensionsFromOptions(context, options);
            widget.setDimensions(dims);
            updateAppWidget(manager, widget);
        }
        catch (RuntimeException e)
        {
            drawErrorWidget(context, manager, widgetId, e);
            e.printStackTrace();
        }
    }

    @Override
    public void onDeleted(@Nullable Context context, @Nullable int[] ids)
    {
        if (context == null) throw new RuntimeException("context is null");
        if (ids == null) throw new RuntimeException("ids is null");

        updateDependencies(context);

        for (int id : ids)
        {
            try
            {
                BaseWidget widget = getWidgetFromId(context, id);
                widget.delete();
            }
            catch (HabitNotFoundException e)
            {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onUpdate(@Nullable Context context,
                         @Nullable AppWidgetManager manager,
                         @Nullable int[] widgetIds)
    {
        if (context == null) throw new RuntimeException("context is null");
        if (manager == null) throw new RuntimeException("manager is null");
        if (widgetIds == null) throw new RuntimeException("widgetIds is null");
        updateDependencies(context);
        context.setTheme(R.style.WidgetTheme);

        new Thread(() ->
        {
            Looper.prepare();
            for (int id : widgetIds)
                update(context, manager, id);
        }).start();
    }

    protected List<Habit> getHabitsFromWidgetId(int widgetId)
    {
        long selectedIds[] = widgetPrefs.getHabitIdsFromWidgetId(widgetId);
        ArrayList<Habit> selectedHabits = new ArrayList<>(selectedIds.length);
        for (long id : selectedIds)
        {
            Habit h = habits.getById(id);
            if (h == null) throw new HabitNotFoundException();
            selectedHabits.add(h);
        }

        return selectedHabits;
    }

    @NonNull
    protected abstract BaseWidget getWidgetFromId(@NonNull Context context,
                                                  int id);

    private void drawErrorWidget(Context context,
                                 AppWidgetManager manager,
                                 int widgetId,
                                 RuntimeException e)
    {
        RemoteViews errorView =
            new RemoteViews(context.getPackageName(), R.layout.widget_error);

        if (e instanceof HabitNotFoundException)
        {
            errorView.setCharSequence(R.id.label, "setText",
                context.getString(R.string.habit_not_found));
        }

        manager.updateAppWidget(widgetId, errorView);
    }

    private void update(@NonNull Context context,
                        @NonNull AppWidgetManager manager,
                        int widgetId)
    {
        try
        {
            BaseWidget widget = getWidgetFromId(context, widgetId);
            Bundle options = manager.getAppWidgetOptions(widgetId);
            widget.setDimensions(getDimensionsFromOptions(context, options));
            updateAppWidget(manager, widget);
        }
        catch (RuntimeException e)
        {
            drawErrorWidget(context, manager, widgetId, e);
            e.printStackTrace();
        }
    }

    private void updateDependencies(Context context)
    {
        HabitsApplication app =
            (HabitsApplication) context.getApplicationContext();
        habits = app.getComponent().getHabitList();
        preferences = app.getComponent().getPreferences();
        widgetPrefs = app.getComponent().getWidgetPreferences();
    }

    public Preferences getPreferences()
    {
        return preferences;
    }
}

File: uhabits-android/src/main/java/org/isoron/uhabits/widgets/BaseWidgetProvider.kt
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.isoron.uhabits.widgets

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.os.Bundle
import android.os.Looper
import android.widget.RemoteViews
import org.isoron.uhabits.HabitsApplication
import org.isoron.uhabits.R
import org.isoron.uhabits.core.models.Habit
import org.isoron.uhabits.core.models.HabitList
import org.isoron.uhabits.core.models.HabitNotFoundException
import org.isoron.uhabits.core.preferences.Preferences
import org.isoron.uhabits.core.preferences.WidgetPreferences
import org.isoron.uhabits.utils.InterfaceUtils.dpToPixels
import java.util.ArrayList

abstract class BaseWidgetProvider : AppWidgetProvider() {
    private lateinit var habits: HabitList
    lateinit var preferences: Preferences
        private set
    private lateinit var widgetPrefs: WidgetPreferences
    fun getDimensionsFromOptions(
        ctx: Context,
        options: Bundle
    ): WidgetDimensions {
        val maxWidth = dpToPixels(
            ctx,
            options.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH).toFloat()
        ).toInt()
        val maxHeight = dpToPixels(
            ctx,
            options.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_HEIGHT).toFloat()
        ).toInt()
        val minWidth = dpToPixels(
            ctx,
            options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH).toFloat()
        ).toInt()
        val minHeight = dpToPixels(
            ctx,
            options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT).toFloat()
        ).toInt()
        return WidgetDimensions(minWidth, maxHeight, maxWidth, minHeight)
    }

    override fun onAppWidgetOptionsChanged(
        context: Context,
        manager: AppWidgetManager,
        widgetId: Int,
        options: Bundle
    ) {
        try {
            updateDependencies(context)
            context.setTheme(R.style.WidgetTheme)
            val widget = getWidgetFromId(context, widgetId)
            val dims = getDimensionsFromOptions(context, options)
            widget.setDimensions(dims)
            updateAppWidget(manager, widget)
        } catch (e: RuntimeException) {
            drawErrorWidget(context, manager, widgetId, e)
            e.printStackTrace()
        }
    }

    override fun onDeleted(context: Context?, ids: IntArray?) {
        if (context == null) throw RuntimeException("context is null")
        if (ids == null) throw RuntimeException("ids is null")
        updateDependencies(context)
        for (id in ids) {
            try {
                val widget = getWidgetFromId(context, id)
                widget.delete()
            } catch (e: HabitNotFoundException) {
                e.printStackTrace()
            }
        }
    }

    override fun onUpdate(
        context: Context,
        manager: AppWidgetManager,
        widgetIds: IntArray
    ) {
        updateDependencies(context)
        context.setTheme(R.style.WidgetTheme)
        Thread {
            Looper.prepare()
            for (id in widgetIds) update(context, manager, id)
        }.start()
    }

    protected fun getHabitsFromWidgetId(widgetId: Int): List<Habit> {
        val selectedIds = widgetPrefs.getHabitIdsFromWidgetId(widgetId)
        val selectedHabits = ArrayList<Habit>(selectedIds.size)
        for (id in selectedIds) {
            val h = habits.getById(id) ?: throw HabitNotFoundException()
            selectedHabits.add(h)
        }
        return selectedHabits
    }

    protected abstract fun getWidgetFromId(
        context: Context,
        id: Int
    ): BaseWidget

    private fun drawErrorWidget(
        context: Context,
        manager: AppWidgetManager,
        widgetId: Int,
        e: RuntimeException
    ) {
        val errorView = RemoteViews(context.packageName, R.layout.widget_error)
        if (e is HabitNotFoundException) {
            errorView.setCharSequence(
                R.id.label,
                "setText",
                context.getString(R.string.habit_not_found)
            )
        }
        manager.updateAppWidget(widgetId, errorView)
    }

    private fun update(
        context: Context,
        manager: AppWidgetManager,
        widgetId: Int
    ) {
        try {
            val widget = getWidgetFromId(context, widgetId)
            val options = manager.getAppWidgetOptions(widgetId)
            widget.setDimensions(getDimensionsFromOptions(context, options))
            updateAppWidget(manager, widget)
        } catch (e: RuntimeException) {
            drawErrorWidget(context, manager, widgetId, e)
            e.printStackTrace()
        }
    }

    private fun updateDependencies(context: Context) {
        val app = context.applicationContext as HabitsApplication
        habits = app.component.habitList
        preferences = app.component.preferences
        widgetPrefs = app.component.widgetPreferences
    }

    companion object {
        fun updateAppWidget(
            manager: AppWidgetManager,
            widget: BaseWidget
        ) {
            val landscape = widget.landscapeRemoteViews
            val portrait = widget.portraitRemoteViews
            val views = RemoteViews(landscape, portrait)
            manager.updateAppWidget(widget.id, views)
        }
    }
}
